NMS-14617: Quick Start: Set up a Threshold (#5268)
OpenNMS/opennms: docs/modules/operation/pages/deep-dive/thresholds/datasource.adoc

This procedure uses metadata to trigger an event when the number of logged-in users exceeds one.

The expression is in the form `${context:key|context_fallback:key_fallback|...|default}`.

Before using metadata in a threshold, you need to add the metadata context pair—in this case, a requisition key called `userLimit` (see <<deep-dive/meta-data.adoc#ga-metadata-webui, Adding metadata through the web UI>>).

. Click the *Gear* symbol in the top-right of the screen.
. Under Performance Measurement, click *Configure Thresholds*.
. Click *Edit* beside the `netsnmp` group.
. Click *Create New Expression-based Threshold*.
. Fill in the following information:
+
* Type: High
* Expression: `hrSystemNumUsers / ${requisition:userLimit|1}`
* Datasource type: Node
* Value: 1
* Rearm: 1
* Description: Too many logged-in users
+
image::metadata/meta-expression1.png[]

. Click *Save*.

This expression will trigger an event when the number of logged-in users exceeds one.

image::metadata/meta-expression2.png[]

[[metric-collect]]
== Confirm that you are collecting metrics

Before creating a threshold, you should make sure you are collecting the metric against which you want to threshold:

> . In the {page-component-title} UI, choose menu:Reports[Resource Graphs].

Reviewer: ```suggestion
. Choose menu:Reports[Resource Graphs].
```

. Select one of the listed resources.

> . Under *SNMP Node Data*, select menu:Node-Level Performance Data[Graph Selection].

Reviewer: ```suggestion
. Under SNMP Node Data, select menu:Node-Level Performance Data[Graph Selection].
```

. Scroll to find the graph for the metric you want to threshold.
+
You can click the *Binoculars* symbol to display only that graph.

[[datasource-determine]]
== Determine the data source

Creating a threshold requires the name of the data source generating the metrics on which you want to threshold.
Data source names for the SNMP protocol appear in `etc/snmp-graph.properties.d`.

To determine the name of the data source, navigate to the *Resource Graphs* screen:

. Choose menu:Reports[Resource Graphs].
. Select one of the listed resources.

> . Under *SNMP Node Data*, select menu:Node-Level Performance Data[Graph Selection].

Reviewer: ```suggestion
. Under SNMP Node Data, select menu:Node-Level Performance Data[Graph Selection].
```

. Scroll through the graphs to find the title of the graph that displays the metric on which you want to threshold.
For example, "Number of Processes" or "System Uptime":
+
image::thresholding/Graphs.png[]

. Go to `etc/snmp-graph.properties.d` and search for the title of the graph (for example, "System Uptime").

> . Note the name of the data source, and type it in the *Datasource* field when you xref:quick-start/thresholding.adoc[create your threshold].

Reviewer: ```suggestion
. Note the name of the data source, and type it in the *Datasource* box when you xref:quick-start/thresholding.adoc[create your threshold].
```

= Threshold Creation

For information on how to create a basic or expression-based threshold, see <<quick-start/thresholding.adoc#setup-threshold, Set up a threshold>> in the Quick Start section.
This section describes related thresholding tasks.

[[ga-threshold-metadata]]
== Use metadata in a threshold

Metadata in expression-based thresholds can streamline threshold creation.
The <<deep-dive/meta-data.adoc#ga-meta-data-dsl, Metadata DSL>> (domain specific language) lets you use patterns in an expression, whereby the metadata is replaced with a corresponding value during the collection process.

> A single expression can behave differently based on the node being tested against.

Reviewer: ```suggestion
A single expression can behave differently based on the node it is tested against.
```

During evaluation of an expression, the following scopes are available:

* Node metadata
* Interface metadata
* Service metadata

Metadata is also supported in value, re-arm, and trigger fields for Single-DS and expression-based thresholds.
For more information on metadata and how to define it, see <<deep-dive/meta-data.adoc#metadata-overview, Metadata>>.

=== Example

> This procedure uses metadata to trigger an event when the number of logged-in users exceeds 1.

Reviewer: ```suggestion
This procedure uses metadata to trigger an event when the number of logged-in users exceeds one.
```

The expression is in the form `${context:key|context_fallback:key_fallback|...|default}`.

Before using metadata in a threshold, you need to add the metadata context pair—in this case, a requisition key called `userLimit` (see <<deep-dive/meta-data.adoc#ga-metadata-webui, Adding metadata through the web UI>>).

. Click the *Gear* symbol in the top-right of the screen.

> . Under *Performance Measurement*, choose *Configure Thresholds*.

Reviewer: ```suggestion
. Under Performance Measurement, click *Configure Thresholds*.
```

. Click *Edit* beside the `netsnmp` group.
. Click *Create New Expression-based Threshold*.
. Fill in the following information:
+
* Type: High
* Expression: `hrSystemNumUsers / ${requisition:userLimit|1}`
* Datasource type: Node
* Value: 1
* Rearm: 1
* Description: Too many logged-in users
+
image::metadata/meta-expression1.png[]

. Click *Save*.

> This expression will trigger an event when the number of logged-in users exceeds 1.

Reviewer: ```suggestion
This expression will trigger an event when the number of logged-in users exceeds one.
```

image::metadata/meta-expression2.png[]

[[metric-collect]]
== Determine that you are collecting the metric

Before creating a threshold, you should make sure you are collecting the metric against which you want to threshold:

. In the {page-component-title} UI, choose menu:Reports[Resource Graphs].
. Select one of the listed resources.
. Under *SNMP Node Data*, select menu:Node-Level Performance Data[Graph Selection].
. Scroll to find the graph for the metric you want to threshold.
+

> You can click the binoculars icon to display only that graph.

Reviewer: ```suggestion
You can click the *Binoculars* symbol to display only that graph.
```

[[datasource-determine]]
== Determine the data source

Creating a threshold requires the name of the data source generating the metrics on which you want to threshold.
Data source names for the SNMP protocol appear in `etc/snmp-graph.properties.d/`.

To determine the name of the data source, navigate to the *Resource Graphs* screen:

. Choose menu:Reports[Resource Graphs].
. Select one of the listed resources.
. Under *SNMP Node Data*, select menu:Node-Level Performance Data[Graph Selection].
. Scroll through the graphs to find the title of the graph that displays the metric on which you want to threshold.
For example, "Number of Processes" or "System Uptime":
+
image::thresholding/Graphs.png[]

> . Go to `etc/snmp-graph.properties.d/` and search for the title of the graph (for example, "System Uptime").

Reviewer: ```suggestion
. Go to `etc/snmp-graph.properties.d` and search for the title of the graph (for example, "System Uptime").
```

> 
> . Note the name of the data source, and type it in the *Datasource* field when you xref:quick-start/thresholding.adoc[create your threshold].

Reviewer: ```suggestion
. Note the name of the data source, and type it in the *Datasource* field when you xref:quick-start/thresholding.adoc[create your threshold].
```
